### PR TITLE
Add references for all methods

### PR DESCRIPTION
**Pull request by amontoison**
@dpo, do you have a publication associated to 
```
D. Orban,
The Projected Golub-Kahan Process for Constrained Linear Least-Squares Problems,
Cahier du GERAD G-2014-15, 2014.
```
?

### CODE REVIEW

**Comment by dpo** [comment]
It's not published. I've been meaning to revise it for a long time.

File: src/cg_lanczos.jl [hunk 1]
```diff
@@ -26,6 +25,10 @@ The method does _not_ abort if A is not definite.
 
 A preconditioner M may be provided in the form of a linear operator and is
 assumed to be symmetric and positive definite.
+
+#### Reference
+
+* A. Frommer and P. Maass, *Fast CG-Based Methods for Tikhonov-Phillips Regularization*, SIAM Journal on Scientific Computing, 20(5), pp. 1831--1850, 1999.
```
Comment (dpo) [inline src/cg_lanczos.jl:31]: Paige and Saunders (1975) describe CG_Lanczos (the same paper as MINRES and SYMMLQ).

File: src/cgls.jl [hunk 2]
```diff
@@ -41,6 +43,10 @@ but is more stable.
 CGLS produces monotonic residuals ‖r‖₂ but not optimality residuals ‖Aᵀr‖₂.
 It is formally equivalent to LSQR, though can be slightly less accurate,
 but simpler to implement.
+
+#### Reference
+
+* A. Björck, T. Elfving and Z. Strakos, *Stability of Conjugate Gradient and Lanczos Methods for Linear Least Squares Problems*, SIAM Journal on Matrix Analysis and Applications, 19(3), pp. 720--736, 1998.
```
Comment (dpo) [inline src/cgls.jl:49]: Hestenes and Stiefel already describe CGLS in the 1952 opus.

File: src/cgs.jl [hunk 3]
```diff
@@ -36,6 +36,11 @@ to become inaccurate.
 TFQMR and BICGSTAB were developed to remedy this difficulty.»
 
 This implementation allows a left preconditioner M and a right preconditioner N.
+
+#### References
+
+* Y. Saad, *Iterative methods for sparse linear systems*, PWS Publishing Company, Boston, USA, 1996.
```
Comment (dpo) [inline src/cgs.jl:42]: I don't think citing books really helps. Why cite one book and not another?

File: src/craigmr.jl [hunk 4]
```diff
@@ -64,6 +63,11 @@ CRAIGMR produces monotonic residuals ‖r‖₂.
 It is formally equivalent to CRMR, though can be slightly more accurate,
 and intricate to implement. Both the x- and y-parts of the solution are
 returned.
+
+#### References
+
+* D. Orban and M. Arioli. *Iterative Solution of Symmetric Quasi-Definite Linear Systems*, Volume 3 of Spotlights. SIAM, Philadelphia, PA, 2017.
```
Comment (dpo) [inline src/craigmr.jl:69]: Of course in this book, we derive CRAIGMR :-)

**Comment by amontoison** [comment]
> It's not published. I've been meaning to revise it for a long time.

With `LNLQ`, `LSLQ` and some numerical tests it could be even better.
But I suppose that you already know and the limiting factor is time.

**Comment by dpo** [comment]
Yes I've been meaning for a long time to fix the SYMMLQ explanation in light of LSLQ and LNLQ. Time is too short. I should do it though.

**Pull request merged**